### DISCUSSION TITLE
2D axisymmetric FFD mesh deformation in two directions.

### DISCUSSION THREAD

**Discussion by ChristianBauerEng** [su2code/SU2#2029] [Q&A]
Dear all,

I'd like to use SU2 for automatic shape deformation of a cavity. My use case is similar to the "Unconstrained shape design of a two way mixing channel" tutorial. I could successfully reproduce the deformation in vertical direction. Now I'd additionally like to optimize the length of the mixing channel outlet pipe.

My idea to do this was to perform two deformations in each optimization step:
- Deform wall in vertical direction
- Deform wall and axis in horizontal direction

I hope this sketch makes clearer what I'm trying to achieve:
![grafik](https://github.com/su2code/SU2/assets/42602495/be519ee6-d581-4a9e-907a-b0bd66f39f94)

In the lower halve you can see the original, undeformed mesh and in red the corresponding undeformed FFD-box. The upper halve shows the deformed mesh after  vertical deformation and in white the deformed FFD-Box. So far it looks good.

However, when further deforming the mesh in horizontal direction it breaks the mesh:
![grafik](https://github.com/su2code/SU2/assets/42602495/bf5b864d-23d7-4cf1-b760-1b04b8f2866d)

It looks like the deformation in horizontal direction was indeed performed, but some nodes near the axis seem to be treated differently, breaking the mesh. Also when changing the definition of the FFD-box to encompass all right-side nodes after initial, vertical deformation this happens.

My questions are now:
- Is it possible to specify the file-name of the output volumes in the deformation step? (default: volume_deformed_00000.vtu). Could make analysis easier.
- Is my two-step approach viable or overly complicated. Is there an easier approach?
- Is it possible to define two different FFD-boxes for horizontal/vertical deformations?
- What causes the nodes at the right side to break the mesh?

Thank you!

Regards,
Christian

**Comment by pcarruscag** [comment]
You can have vertical and horizontal in the same deformation. You also need a bigger ffd box that includes the entire outlet otherwise the deformation shears that boundary. It needs to pass a little bit the axis to capture the outlet, but just enough (e.g. 1e-6 if possible) otherwise the deformation will break the symmetry.

**Reply by pcarruscag** [reply]
It breaks the symmetry because the vertical displacement would affect the outlet point that is on the axis.

**Reply by ChristianBauerEng** [reply]
> You can have vertical and horizontal in the same deformation. 

Thank you, I'll give it a try! I'll change the strategy then. I'll set the near-axis FFD-Points close to the axis and will fix their movement in horizontal direction. I'll then calculate their dv-value from the vertical/horizontal movement of the far-axis points to make sure, that the mesh is deformed nicely.

**Reply by ChristianBauerEng** [reply]
So I've changed the FFD-Box definition to encompass the axis (y=-1e-6) and only be moved in horizontal direction.

See again, undeformed FFD-Box/Mesh on the bottom, deformed entities on top:
![grafik](https://github.com/su2code/SU2/assets/42602495/c78cc7a8-dc9a-44c8-bd19-d8c9433b19ba)

A close-up look at the rightmost nodes near the axis:
![grafik](https://github.com/su2code/SU2/assets/42602495/7edc708c-57e3-49c0-a335-e7e96eb56d90)


Still looks like the axis-node is treated differently than the other nodes at the wall on the right. Any idea, what's going on here?

**Reply by pcarruscag** [reply]
Is the bottom boundary set as symmetry or wall? Make sure it is symmetry, or as a last resort include it in the dv markers

**Reply by ChristianBauerEng** [reply]
The lower boundary is an axis. I assumed this would be enough to allow it some movement, but that does not seem to be the case.

I included it now in the dv markers, and now it looks like it should:
![grafik](https://github.com/su2code/SU2/assets/42602495/22e56eac-6838-46d3-a0bd-d2ec33ae5432)

I'm happy with the result now, but the behaviour is a bit unintuitive. Is this working as intended? In the tutorial case the axis nodes move, even though they are not in the dv-marker list.

**Reply by pcarruscag** [reply]
It's hard to say without the config.

**Comment by ChristianBauerEng** [comment]
Turns out this issue is still not resolved. To maybe dive a bit deeper into the topic, here is an overview of the domain topology:

![grafik](https://github.com/su2code/SU2/assets/42602495/5db337a3-8fba-46fb-9f75-786fc38709f2)

As suggested previously I've set the lower FFD-Box points to a vertical coordinate of -1e-6 to enclose them in the FFD-Box. If I now include only the axis near the tip (NS_ResonatorTip_Axis) in the DV_MARKER list, the mesh is torn apart at the contact point of both axis segments:
![grafik](https://github.com/su2code/SU2/assets/42602495/491e21f6-a024-407d-8065-9e6d530c4385)

If I include also the axis segment further to the left (NS_Resonator_Axis) in the DV_MARKER list, then I get the following errors when I add the FFD-information to the mesh:
```
----------------- FFD technique (cartesian -> parametric) ---------------
Unknown point: (0.0120748, 0, 0). Increase the value of FFD_ITERATIONS.
Please check this point: Local (2.30829 0.753241 0.504737) <-> Global (0.0120748 0 0) <-> Error 0.115343 vs 1e-10.
Unknown point: (0.0121568, 0, 0). Increase the value of FFD_ITERATIONS.
Please check this point: Local (1.76776 0.726314 0.519616) <-> Global (0.0121568 0 0) <-> Error 0.0882403 vs 1e-10.
Unknown point: (0.0122386, 0, 0). Increase the value of FFD_ITERATIONS.
Please check this point: Local (2.51086 0.240484 0.497941) <-> Global (0.0122386 0 0) <-> Error 0.125305 vs 1e-10.
```
When I then deform the mesh some nodes at the axis are moved seemingly randomly around, even if the enclosing FFD-Box vertices are kept constant.
![grafik](https://github.com/su2code/SU2/assets/42602495/e3a57a89-cdcb-4438-be84-8e370dce589f)

The unmapped axis-points seem to always occur within the first FFD-Subbox, also when the overall FFD-boundaries are moved to the left or right.

As a workaround I tried to move the FFD-Box point on the lower left slightly above the axis (+1e-6), while the right-most remains below the axis (-1e-6), which at first appeared to yield the desired result. The messages about "Unknown points" during FFD-definition are gone, and the deformed mesh looks nice:
![grafik](https://github.com/su2code/SU2/assets/42602495/fa0cb966-64b0-4edb-af8d-df12885fd258)

However, when starting the actual simulation the solver diverges within the first time-iteration, even though there doesn't seem to be an obvious defect in the mesh.

Here are the input-files used to obtain these results:
[Inputs.zip](https://github.com/su2code/SU2/files/11555546/Inputs.zip)

I think if the issue with the "Unknown Points" can be resolved there's a good chance that calculations on the deformed mesh succeed.

**Reply by pcarruscag** [reply]
Try giving it more room below the axis than 1e-6 e.g. -1e-5, etc. you have more points blowing up than the 3 in the message. The ones in the message are probably on the left of the box, they seem closer together than the problem regions in the picture.

**Reply by ChristianBauerEng** [reply]
Actually the log shows many more Unknown Points, but I didn't copy the full list as I thought it doesn't really add more information.

I've allready tried moving the lower Box points further down, but this also doesn't seem to help.

In order to check if it's a mesh issue I've also moved the FFD-Box further to the right, enclosing the area where there are no "Unknown Points", but this simply generates new Unknown Points in the previously working region. These points seem to always occur in the sub-box with i-index 0.

**Reply by ChristianBauerEng** [reply]
Another issue came up...

It appears that when an FFD-element is deformed so much that it lies entirely or partly on the other side of the axis then the symmetry boundary condition is violated:
![grafik](https://github.com/su2code/SU2/assets/42602495/b9171b1f-c6bf-4175-9475-372e7c1a6b77)

![grafik](https://github.com/su2code/SU2/assets/42602495/b4801ed3-be86-4123-9b7f-6a795821a2ce)
Again, the undistorted mesh is on the bottom, while the deformed mesh is on top. It can be seen, that some nodes now lie below the axis.

Apparently this happens when STIFFNESS_TYPE is set to WALL_DISTANCE, but not to CONSTANT_STIFFNESS.

**Reply by pcarruscag** [reply]
I can't tell from the picture but maybe the points that come under the axis are internal points that overshoot due to excessive deformation?
I'll try to see what is going on with the point mapping when I have some time.
